docs: Cleanup & Standardize Guides (#1357)
ai-dynamo/dynamo: docs/guides/disagg_perf_tuning.md

# Disaggregation and Performance Tuning

Disaggregation gains performance by separating the prefill and decode into different engines to reduce interferences between the two.
However, performant disaggregation requires careful tuning of the inference parameters.
Specifically, there are three sets of parameters that needs to be tuned:

  1. Engine configuration and options (e.g. parallelization mapping, maximum number of tokens, etc.).
  2. Disaggregated router configuration and options.
  3. Number of prefill and decode engines.

This guide describes the process of tuning these parameters.

## Engine Tuning

The most important engine knob to tune is the parallelization mapping.
For most dense models, the best setting is to use TP within node and PP across nodes.
For example, for llama 405b w8a8 on H100, TP8 on a single node or TP8PP2 on two nodes is usually the best choice.

> The next thing to decide is how many numbers of GPU to serve the model.

Reviewer: ```suggestion
The next thing to decide is how many GPUs to use to serve the model.
```

Author: Fixed up.

Author: @saturley-hall resolve?

Typically, the number of GPUs vs the performance follows the following pattern:

| Number of GPUs                                      | Performance
| :-------------------------------------------------- | :---------------------------------------------------------------------------------------- |
| Cannot hold weights in VRAM                         | OOM                                                                                       |

> | (Barely hold weights in VRAM)                       | (KV cache is too small to maintain large enough sequence length or reasonable batch size) |

Reviewer: Is the parenthetical notation necessary here?

Author: @saturley-hall resolve?

| Minimum number with fair amount of KV cache         | Best overall throughput/GPU, worst latency/user                                           |
| Between minimum and maximum                         | Tradeoff between throughput/GPU and latency/user                                          |
| Maximum number limited by communication scalability | Worst overall throughput/GPU, best latency/user                                           |
| More than maximum                                   | Communication overhead dominates, poor performance                                        |

> Note that for decode-only engines, sometimes larger number of GPUs has to larger KV cache per GPU and more decoding requests running in parallel, which leads to both better throughput/GPU and better latency/user.

Reviewer: I am not sure what this sentence is saying

Author: neither do I, again not my prose. I was just reformatting for readability.

For example, for Llama-3.3-70b NVFP4 quantization on B200 in vLLM with 0.9 free GPU memory fraction:

| TP Size | KV Cache Size (GB) | KV Cache per GPU (GB) | Per GPU Improvement over TP1 |
| ------: | -----------------: | --------------------: | ---------------------------: |
|       1 |                113 |                   113 |                        1.00x |
|       2 |                269 |                   135 |                        1.19x |
|       4 |                578 |                   144 |                        1.28x |

> The best number of GPUs to use in the prefill and decode engines can be determined by running a few fixed isl/osl/concurrency test using [GenAI-Perf](https://github.com/triton-inference-server/perf_analyzer/tree/main/genai-perf) and compare with the SLA.

Reviewer: ```suggestion
The best number of GPUs to use in the prefill and decode engines can be determined by running a few fixed isl/osl/concurrency tests using [GenAI-Perf](https://github.com/triton-inference-server/perf_analyzer/tree/main/genai-perf) and compare with the SLA.
```

Author: Fixed up.

Author: @saturley-hall resolve?

GenAI-Perf is pre-installed in the dynamo container.

Besides the parallelization mapping, other common knobs to tune are maximum batch size, maximum number of tokens, and block size.

> For prefill engines, usually a small batch size and large max_num_token is preferred.

Reviewer: ```suggestion
For prefill engines, usually a small batch size and large `max_num_token` is preferred.
```

Author: Fixed up.

Author: @saturley-hall resolve?

> For decode engines, usually a large batch size and medium max_num_token is preferred.

Reviewer: ```suggestion
For decode engines, usually a large batch size and medium `max_num_token` is preferred.
```

Author: @saturley-hall resolve?

> For details on tuning the max_num_token and max_batch_size, see the next section.

Reviewer: ```suggestion
For details on tuning the `max_num_token` and `max_batch_size`, see the next section.
```

Author: Fixed up.

Hence, the current best strategy is to set the maximum batch size to the optimized KV cache size and set the maximum number of tokens to the maximum local prefill length + maximum batch size (since one decode request has one active token).

## Number of Prefill and Decode Engines

The best dynamo knob choices depends on the operating condition of the model.
Based on the load, we define three operating conditions:

  1. **Low load**:
     The endpoint is hit by a single user (single-stream) most of the time.

  2. **Medium load**:
     The endpoint is hit by multiple users, but the KV cache of the decode engines is never fully utilized.

  3. **High load**:
     The endpoint is hit by multiple users and the requests are queued up due to no available KV cache in the decode engines.

At low load, disaggregation would not benefit much as prefill and decode are usually computed separately.
It is usually better to use a single monolithic engine.

At medium load, disaggregation allows better ITL compared with prefill-prioritized and chunked prefill engines and better TTFT compared with chunked prefill engine and decode-only engine for each user.

> Dynamo users can adjust the number of prefill and decode engines based on TTFT and ITL SLA.

Reviewer: Should we call them `users` or `operators`?
